### DISCUSSION TITLE
how to keep_alive when returning a list or tuple?

### DISCUSSION THREAD

**Discussion by wojdyr** [wjakob/nanobind#816] [Q&A] [answered]
Let's say that I have `Item`s that reference `Parent`.

```c++
struct Item {
    Parent* parent;
};
```

`Parent` can return a vector with a number of such items 
```c++
struct Parent {
    std::vector<Item> get_items() {
        return {Item{this}, Item{this}};
    }
};
```

All items should nurse their parent. If a single `Item` was returned, I'd just add `keep_alive`:
```c++
.def("get_single_item", &Parent::get_single_item, nb::keep_alive<0, 1>())
```

but if I try to do the same with a vector of items (the function `get_items()` defined above):
```c++
.def("get_items", &Parent::get_items, nb::keep_alive<0, 1>())
```
it complains that `nanobind::detail::keep_alive(): could not create a weak reference! Likely, the 'nurse' argument you specified is not a weak-referenceable type!`. (OK, the list can't be a nurse.)

How to solve it?

If I could use `rv_policy::reference_internal`, the policy would be passed to all elements and I think everything would just work. But in my case the items are not internal members. Is there a way to somehow pass `keep_alive` to all elements of a list?

**Comment by wjakob** [comment, marked as answer]
You're in luck: @oremanj added a new feature in the latest version of nanobind that I think could be used to support your use case: https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I0EN8nanobind11call_policyE

**Reply by wojdyr** [reply]
Great! Thanks @oremanj for solving my problem before I asked.
The example in docs is exactly what I needed, with a minor fix: `I > 0 && I < N` seems to be off by one. I think it meant to be `<= N` and then `args[I-1]` when calling `detail::keep_alive`.
